### ISSUE TITLE
gopherjs can no longer be used from go generate

### ISSUE THREAD

**Issue by DrJosh9000** [gopherjs/gopherjs#601]
Setup:
$ cat generate.go
//go:generate gopherjs build
//(other go:generate stuff using the .js output)

Expected behaviour:
Running go generate runs gopherjs, which builds the .js and .js.map as if running gopherjs build directly.

Actual output:
gopherjs: unsupported GOOS/GOARCH pair darwin/amd64

go generate sets $GOOS and $GOARCH (https://golang.org/src/cmd/go/generate.go).
go generate has no way of overriding $GOOS or $GOARCH from the generate directive: //go:generate GOARCH=js gopherjs build just makes go generate try to exec "GOARCH=js".

**Comment by shurcooL** [comment]
Indeed, this is a valid unintended bug, a consequence of resolving #594.

go generate sets some environment variables, including GOARCH to the current architecture.

This issue is identical to what one would run into if they tried to cross-compile from a go generate directive.

First idea that comes to my mind is `env` command:

```
ENV(1)                    BSD General Commands Manual                   ENV(1)

NAME
     env -- set environment and execute command, or print environment

SYNOPSIS
     env [-iv] [-P altpath] [-S string] [-u name] [name=value ...]
         [utility [argument ...]]
```

Would this work?

```
//go:generate env GOARCH=js gopherjs build
```

**Comment by DrJosh9000** [comment]
env does the trick--thanks! Seems like this should instead be a hint in the documentation for go generate.

**Comment by shurcooL** [comment]
> env does the trick--thanks!

Glad to hear, great!

> Seems like this should instead be a hint in the documentation for go generate.

Which documentation are you referring to? Can you link me to it please?

**Comment by DrJosh9000** [comment]
I'm thinking of this:
https://golang.org/cmd/go/#hdr-Generate_Go_files_by_processing_source

**Comment by shurcooL** [comment]
I see.

I'm not completely sure it's worth mentioning there, because that is the actual specification of the go generate command and `//go:generate` directives, not a tutorial. `env` is a Unix-specific command that Windows or other OSes probably don't have. However, perhaps there's another place that would be a better fit (FAQ, or some tutorial or user guide that covers additional things related to `go generate`). If you think that the current situation can be improved, I suggest opening an issue in the Go issue tracker.

I'll close this issue, since there's nothing more for us to do in this GopherJS specific issue. Feel free to comment otherwise. Thanks for reporting this! :)

**Comment by shurcooL** [comment]
Reopening this issue.

We've originally discussed two possible fixes in issue #594:

> The `gopherjs` tool does not use the `GOARCH` environment variable in any way. Internally, GopherJS always uses `js` as `GOARCH` value when compiling user code, but it does other internal stuff when compiling standard library.
>
> 1. One fix is to make the `gopherjs` tool really ignore `GOARCH` env var. In other words, setting it to any value should have no effect; it shouldn't cause this failure.
>
> 2. We could also consider making it use the value of `GOARCH`, and error out if it's set to anything other than `js` (which is the default and the only supported value). Since `gopherjs` can't compile for any other architectures.

Upon further consideration, after seeing this issue #601 and gopherjs/gopherjs.github.io#66, we've decided (see discussion in https://github.com/gopherjs/gopherjs.github.io/pull/66#issuecomment-286519322) to revisit this issue and change gopherjs to completely ignore `GOARCH` environment variable, instead of requiring it to be unset or equal to `js`.

I'll make a PR for that shortly. Reopening this issue to track that.
